### PR TITLE
feat(wandb): add sync_step

### LINKED ISSUE #4980: wandb logger problem with on_step log on validation
## 🐛 Bug

When logging on the `validation_step` with `on_step=True` and `on_epoch=False` the following happens:

- wandb warnings are generated to alert about a step numbering problem (probably confusing the validation step number which seems cyclical with the overall step which is always increasing)

image

- wandb charts *for training* (by step) is shrunk on the x dimension (like the number of steps for the whole training were less). We tested 2 training runs: the first (blue in the image below) with  `on_step=False` and `on_epoch=True` on `validation_step`, the second with `on_step=True` and `on_epoch=False` (red in the image below). As you can see the *training* chart is affected by this:

image

- an error is issued at the end of the second training run:

image

- two new (unrequested) panels appear at the top to the wandb project (this is the weirdest of the lot :-))

image

### To Reproduce

Just change the `validation_step` logging like this: 

    def validation_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        loss = F.nll_loss(logits, y)

        # validation metrics
        preds = torch.argmax(logits, dim=1)
        acc = accuracy(preds, y)
        self.log('val_loss', loss, on_step=True, on_epoch=False, prog_bar=True)
        self.log('val_acc', acc, on_step=True, on_epoch=False, prog_bar=True)
        return loss

### LINKED ISSUE #5070: Logging validation metrics with `val_check_interval`
#### What is your question?

My training epoch is very long, so I'd like to run mini-validation loops in the middle of the epoch, which I can do easily by setting `val_check_interval=N`. However, I'd also like to log the values outputted by those mini-validation loops.

**Is there a canonical way to log validation metrics *within* a training epoch while using `val_check_interval` in the `Trainer`?** 

#### Related issues:
https://github.com/PyTorchLightning/pytorch-lightning/issues/4980 -- However, I'm not trying to log step-wise validation metrics; I'm trying to log aggregate validation epoch metrics.
https://github.com/PyTorchLightning/pytorch-lightning/issues/4985 -- Similar, but I'm using the `WandbLogger` directly while this issue is more general.
https://github.com/PyTorchLightning/pytorch-lightning/issues/4652 -- Similar issue, but this is an older version of pytorch-lightning.

#### Code

The below is my code for the methods that involve logging. Unfortunately, I can't post the full repo; here's the relevant configuration (loaded from YAML) for the `Trainer` and `Logger` objects, which I've replicated in a Colab in the below comment.
```
trainer:
    gradient_clip_val: 0.1
    accumulate_grad_batches: 8
    num_sanity_val_steps: 2
    gpus: 1
    max_epochs: 1
    val_check_interval: 2000
    limit_val_batches: 100
    move_metrics_to_cpu: True
logging:
    save_dir: "sentcomp_logs/"
    project: "sent-comp"
    offline: False
    name: "bart_full_run"
```

The following is the actual logging code. For context, I am training a Seq2Seq model from HuggingFace.

```
# THIS WORKS FINE!
    def training_step(self, batch: BatchEncoding, batch_idx: int) -> torch.Tensor:
        outputs = self(**batch)
        loss = outputs[0]
        self.log("train_loss", loss, prog_bar=True, sync_dist=True)
        return loss

# DOES NOT WORK
    def validation_step(self, batch: BatchEncoding, batch_idx: int) -> Dict[str, torch.Tensor]:
        outputs = self(**batch)
        val_loss = outputs[0]
        predictions, targets = self.decode(batch) # returns lists of strings
        results = self.score(predictions, targets) # returns a dictionary
        self.log("val_loss", val_loss, sync_dist=True)
        self.log_dict(results, sync_dist=True)
        return results
```

**Expected:** Logs to my `wandb` project.
**Actual:** A warning `wandb: WARNING Step must only increase in log calls`. 

It seems like the global step of  the validator is out of sync with my logging calls, as I get a message like 
```
Step 249 < 2000; dropping { <my metrics> }
Step 499 < 4000; dropping { <my metrics> }
Step 749 < 6000; dropping { <my metrics> }
```
The global step numbers are definitely consistent with the `val_check_interval` provided (`2000 * n`). However, I am unsure why the validation logging step results in an attempt to log to steps in increments of 250. *Maybe I need to somehow sync the validation logging step to the training logging step?* Or is pytorch lightning supposed to take care of this interally?

Is this a bug, or am I doing something wrong/silly?

#### What have you tried?

Removing the `log_dict` call from validation steps and instead logging an aggregated dictionary in `validation_epoch_end`:

```
def validation_epoch_end(self, batches: List[Dict[str, torch.Tensor]]) -> None:
    metrics = {}         
    metrics[metric_key] = torch.stack([batch[metric_key] for batch in batches]).mean()  
    self.log_dict(metrics, sync_dist=True)
```
This does not change the overall behavior.

#### What's your environment?

 - OS: Ubuntu 18.04
 - Packaging: pip
 - Version: pytorch-lightning 1.1.0 (Master), pytorch 1.7.0, cuda 11.0

### LINKED ISSUE #5194: WandbLogger: add more flexibility with logging step
## 🚀 Feature

Let `WandbLogger` log using:
* the trainer step by default (current behavior)
* an auto-incremented step, independent from the trainer

### Motivation

The current default of using trainer step is a good default and seems to work fine for most people, automatically associated correctly training and validation metrics at the right step.

However, some users have a few issues with custom logging:
* #4980 wants to see a chart of validation metrics over a single batch
* #5070 tries to log intermediate validation metrics, which uses a different step
* https://github.com/wandb/client/issues/1626
* https://github.com/wandb/client/issues/1507

#5050 solves some of these issues by suggesting using `wandb.log(my_dict, commit=False)` when logging outside of the "regular" workflow (which will automatically associate metrics to last used step) but not all of them.
In some cases it may be hard to ensure the associated step is always increasing.

### Pitch

We will be able to have a new parameter in `WandbLogger` (TBD) such as `sync_step=True`:
* by default `True`, we will have the current behavior which syncs `wandb.step` with the trainer step.
* when set to `False`, the step will auto-increment at each logging (by relying on `wandb.log()` default behaviour) and the trainer step will just be added to logged metrics when available

In those cases (set to `False`), we will still be able to manually select the trainer step as x-axis in the W&B dashboard (global graphs or any specific graph) so this feature will add more flexibility in logging.

It will not be the default behavior mainly for people not familiar with W&B and that expect to see validation and training metrics aligned by default (see #4113).

### PR DESCRIPTION
**Pull request by borisdayma**
## What does this PR do?

Add more flexiblitiy to `WandbLogger` with `sync_step` parameter.

Fixes #5194
Fixes #4980
Fixes #5070

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes [if needed]?
- [x] Did you write any new necessary tests [no need for typos, docs]? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone are aligned!**
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
this shall go in release/1.2-dev?

**Comment by Borda** [comment]
@borisdayma mind rebase your change on `release/1.2-dev` 🐰

**Comment by borisdayma** [comment]
I tried to rebase, then just merge `release/1.2-dev` but there's a lot of conflicts (on files I didn't touch) so it scares me a little bit…

File: pytorch_lightning/loggers/wandb.py [hunk 1]
```diff
@@ -46,19 +46,20 @@ class WandbLogger(LightningLoggerBase):
...
         offline: Run offline (data can be streamed later to wandb servers).
         id: Sets the version, mainly used to resume a previous run.
+        version: Same as id.
         anonymous: Enables or explicitly disables anonymous logging.
-        version: Sets the version, mainly used to resume a previous run.
         project: The name of the project to which this run will belong.
         log_model: Save checkpoints in wandb dir to upload on W&B servers.
-        experiment: WandB experiment object.
         prefix: A string to put at the beginning of metric keys.
+        sync_step: Sync Trainer step with wandb step.
```
Comment (justusschock) [inline pytorch_lightning/loggers/wandb.py:57]: Can you put this after the experiment flag, so that old args order is still working?

File: pytorch_lightning/loggers/wandb.py [hunk 2]
```diff
@@ -83,14 +84,15 @@ def __init__(
...
+        offline: Optional[bool] = False,
         id: Optional[str] = None,
-        anonymous: bool = False,
         version: Optional[str] = None,
+        anonymous: Optional[bool] = False,
         project: Optional[str] = None,
-        log_model: bool = False,
+        log_model: Optional[bool] = False,
+        prefix: Optional[str] = '',
+        sync_step: Optional[bool] = True,
```
Comment (justusschock) [inline pytorch_lightning/loggers/wandb.py:94]: also put this after `experiment` arg please.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loggers/wandb.py (hunk 1 above)
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:57]: I changed the order within the function itself.
For the description of args, I let this order because I thought it made more sense. For example `id` and `version` are the same thing. Is that ok?

File: pytorch_lightning/loggers/wandb.py [hunk 3]
```diff
@@ -71,9 +72,9 @@ class WandbLogger(LightningLoggerBase):
     make sure to use `commit=False` so the logging step does not increase.
 
     See Also:
-        - `Tutorial <https://app.wandb.ai/cayush/pytorchlightning/reports/
-          Use-Pytorch-Lightning-with-Weights-%26-Biases--Vmlldzo2NjQ1Mw>`__
-          on how to use W&B with Pytorch Lightning.
+        - `Tutorial <https://colab.research.google.com/drive/16d1uctGaw2y9KhGBlINNTsWpmlXdJwRW?usp=sharing>`__
+          on how to use W&B with Pytorch Lightning
```
Comment (awaelchli) [inline pytorch_lightning/loggers/wandb.py:76]:
```suggestion
          on how to use W&B with PyTorch Lightning
```

File: pytorch_lightning/loggers/wandb.py [hunk 4]
```diff
@@ -161,11 +164,16 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
 
         metrics = self._add_prefix(metrics)
-        if step is not None and step + self._step_offset < self.experiment.step:
+        if self._sync_step and step is not None and step + self._step_offset < self.experiment.step:
             self.warning_cache.warn(
-                'Trying to log at a previous step. Use `commit=False` when logging metrics manually.'
-            )
-        self.experiment.log(metrics, step=(step + self._step_offset) if step is not None else None)
+                'Trying to log at a previous step. '
+                'Use `WandbLogger(sync_step=False)` or try logging with `commit=False` when calling manually `wandb.log`.')
```
Comment (awaelchli) [inline pytorch_lightning/loggers/wandb.py:170]:
```suggestion
                'Trying to log at a previous step.'
                ' Use `WandbLogger(sync_step=False)` or try logging with `commit=False` when calling manually `wandb.log`.')
```
just for consistency with rest of code base

File: pytorch_lightning/loggers/wandb.py [hunk 5]
```diff
@@ -161,11 +164,16 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
+        if self._sync_step:
+            self.experiment.log(metrics, step=(step + self._step_offset) if step is not None else None)
+        elif step is not None:
+            self.experiment.log({**metrics, 'trainer_step': (step + self._step_offset)})
```
Comment (awaelchli) [inline pytorch_lightning/loggers/wandb.py:174]: so step is the global_step from trainer, and this trainer_step is the out-of-sync step right?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loggers/wandb.py (hunk 5 above)
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:174]: That is correct. The out-of-sync here is just due to a previous run that is resumed or a second training on a same run.

If you just have one simple training, `trainer_step == global_step` but people used to be confused on what is `global_step` in the past which is why it is renamed a bit more explicitly (+ it will avoid confusion when they are out-of-sync).

If you are talking about the wandb step, you don't see it here (which would be through `step` argument). The fact that it is not given and that there is no `commit=False` means that each batch of metrics will automatically commit and wandb step will increase at each `log`.

Once you are on your dashboard, you can decide to set your x-axis as `trainer_step` which will give you the same graph as if you had not used `sync_step=False`. The main difference is that it will handle cases where you may log multiple points per `global_step` (for example logging each validation batch, etc).

Comment (awaelchli) [reply pytorch_lightning/loggers/wandb.py:174]: great, that make sense. I used to log a step myself and select it in the UI. This will make it easier!

**Comment by borisdayma** [comment]
Hi, just checking if I need to do anything else on my side?
I know of a few people eager to see this PR merged.

File: pytorch_lightning/loggers/wandb.py [hunk 6]
```diff
@@ -163,11 +166,16 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
...
             self.warning_cache.warn(
-                'Trying to log at a previous step. Use `commit=False` when logging metrics manually.'
-            )
-        self.experiment.log(metrics, step=(step + self._step_offset) if step is not None else None)
+                'Trying to log at a previous step. Use `WandbLogger(sync_step=False)`'
+                ' or try logging with `commit=False` when calling manually `wandb.log`.')
+        if self._sync_step:
+            self.experiment.log(metrics, step=(step + self._step_offset) if step is not None else None)
+        elif step is not None:
+            self.experiment.log({**metrics, 'trainer_step': (step + self._step_offset)})
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/wandb.py:176]: why is it done like this?
any reason for not using `self.experiment.log(metrics, step=step + self._step_offset)`

Comment (rohitgr7) [reply pytorch_lightning/loggers/wandb.py:176]: ok ignore this.. saw your comment above.

**Review by rohitgr7 — APPROVED** [review]

**Review by ydcjeff — APPROVED** [review]

File: pytorch_lightning/loggers/wandb.py [hunk 7]
```diff
@@ -85,14 +86,15 @@ def __init__(
         self,
         name: Optional[str] = None,
         save_dir: Optional[str] = None,
-        offline: bool = False,
+        offline: Optional[bool] = False,
```
Comment (akihironitta) [inline pytorch_lightning/loggers/wandb.py:89]: I don't think the change here is necessary. `Optional[bool]` is equivalent to `Union[bool, None]`, and `offline` only accepts `bool`, right?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loggers/wandb.py (hunk 7 above)
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:90]: It should work but actually I want to clean up the way we pass init parameters in a follow-up PR and suggest to just use kwargs.
The possible parameters has evolved and we could just refer to [wandb.init doc](https://docs.wandb.ai/ref/init)

Comment (akihironitta) [reply pytorch_lightning/loggers/wandb.py:90]: That makes sense.
